packages/adapter-d1/src/d1.ts

```diff
@@ -0,0 +1,185 @@
+        columnNames: [],
+        columnTypes: [],
+        rows: [],
+      }
+    }
+
+    const results = ioResult.results as Object[]
+    const columnNames = Object.keys(results[0])
+    const columnTypes = getColumnTypes(columnNames, results)
+    const rows = this.mapD1ToRows(results)
+
+    return {
+      columnNames,
+      // Note: without Object.values the array looks like
+      // columnTypes: [ id: 128 ],
+      // and errors with:
+      // ✘ [ERROR] A hanging Promise was canceled. This happens when the worker runtime is waiting for a Promise from JavaScript to resolve, but has detected that the Promise cannot possibly ever resolve because all code and events related to the Promise's I/O context have already finished.
+      columnTypes: Object.values(columnTypes),
+      rows,
+    }
+  }
+
+  private mapD1ToRows(results: any) {
+    const rows: unknown[][] = []
+    for (const row of results) {
+      const entry = Object.keys(row).map((k) => row[k])
+      rows.push(entry)
+    }
+    return rows
+  }
+
+  /**
+   * Execute a query given as SQL, interpolating the given parameters and
+   * returning the number of affected rows.
+   * Note: Queryable expects a u64, but napi.rs only supports u32.
+   */
+  async executeRaw(query: Query): Promise<Result<number>> {
+    const tag = '[js::execute_raw]'
+    console.debug(`${tag} %O`, query)
+
+    // TODO: rows_written or changes? Only rows_written is documented.
+    return (await this.performIO(query)).map(({ meta }) => meta.changes ?? 0)
+  }
+
+  private async performIO(query: Query): Promise<Result<PerformIOResult>> {
+    // const release = await this[LOCK_TAG].acquire()
+
+    console.debug({ query })
+
+    try {
+      // Hack for
+      // ✘ [ERROR] Error in performIO: Error: D1_TYPE_ERROR: Type 'boolean' not supported for value 'true'
+      query.args = query.args.map((arg) => {
+        if (arg === true) {
+          return 1
+        } else if (arg === false) {
+          return 0
+        }
+        return arg
+      })
+
+      const result = await this.client
+        .prepare(query.sql)
+        .bind(...query.args)
+        .all()
+
+      console.debug({ result })
+
+      return ok(result)
+    } catch (e) {
+      const error = e as Error
+      console.error('Error in performIO: %O', error)
+
+      const rawCode = error['rawCode'] ?? e.cause?.['rawCode']
+      if (typeof rawCode === 'number') {
+        return err({
+          kind: 'Sqlite',
+          extendedCode: rawCode,
+          message: error.message,
+        })
+      }
+      throw error
+    } finally {
+      // release()
+    }
+  }
+}
+
+class D1Transaction extends D1Queryable<StdClient> implements Transaction {
+  constructor(client: StdClient, readonly options: TransactionOptions) {
+    super(client)
+  }
+
+  async commit(): Promise<Result<void>> {
+    console.debug(`[js::commit]`)
+
+    // TODO remove (added to have linting pass)
+    await new Promise((resolve) => resolve)
```

Comment:
Note: this causes wrangler errors
<img width="1233" alt="image" src="https://github.com/prisma/prisma/assets/29753584/e998c70a-e97e-467b-8918-42f2505e0511">